core/include/kernel/notif.h

```diff
@@ -0,0 +1,175 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright (c) 2021, Linaro Limited
+ */
+
+#ifndef __KERNEL_NOTIF_H
+#define __KERNEL_NOTIF_H
+
+#include <compiler.h>
+#include <sys/queue.h>
+#include <tee_api_types.h>
+#include <types_ext.h>
+
+/*
+ * Notification values are divided into two kinds, asynchronous and
+ * synchronous, where the asynchronous has the lowest values.
+ * They are ordered as:
+ * 0			    Do bottom half
+ * 1..NOTIF_ASYNC_MAX	    Free for signalling in PTAs and should be
+ *			    allocated with notif_alloc_async_value()
+ * NOTIF_SYNC_VALUE_BASE..  Used as NOTIF_SYNC_VALUE_BASE + thread_id
+ * NOTIF_VALUE_MAX	    for mutex and condvar wait/wakeup
+ *
+ * Any value can be signalled with notif_send_sync() while only the ones
+ * <= CFG_CORE_ASYNC_NOTIF_MAX can be signalled with notif_send_async().
+ */
+
+#if CFG_CORE_ASYNC_NOTIF_INTID
+#define NOTIF_ASYNC_VALUE_MAX		U(63)
+#define NOTIF_SYNC_VALUE_BASE		(NOTIF_ASYNC_VALUE_MAX + U(1))
+#else
+#define NOTIF_SYNC_VALUE_BASE		0
+#endif
+
+#define NOTIF_VALUE_MAX			(NOTIF_SYNC_VALUE_BASE + \
+					 CFG_NUM_THREADS)
+
+#define NOTIF_VALUE_DO_BOTTOM_HALF	0
+
+/*
+ * enum notif_event - Notification of an event
+ * @NOTIF_EVENT_STARTED:	Delivered in an atomic context to inform
+ *				drivers that normal world has enabled
+ *				asynchronous notifications.
+ * @NOTIF_EVENT_DO_BOTTOM_HALF: Delivered in a yielding context to let a
+ *				driver do bottom half processing.
+ * @NOTIF_EVENT_STOPPED:	Delivered in a yielding contest to inform
+ *				drivers that normal world is about to disable
+ *				asynchronous notifications.
+ *
+ * Once a driver has received a @NOTIF_EVENT_STARTED asynchronous notifications
+ * driving the @NOTIF_EVENT_DO_BOTTOM_HALF deliveries is enabled.
+ *
+ * In case a @NOTIF_EVENT_STOPPED is received there will be no more
+ * @NOTIF_EVENT_DO_BOTTOM_HALF events delivered, until @NOTIF_EVENT_STARTED
+ * has been delivered again.
+ *
+ * Note that while a @NOTIF_EVENT_STOPPED is being delivered at the same
+ * time may a @NOTIF_EVENT_STARTED be delivered again so a driver is
```

Comment:
maybe replace `a driver` with `non-secure world`?
(or did I misunderstood?)

Reply:
A driver is an OP-TEE driver in this case.
I'll remove "shared" from "shared internal state" since I guess that's what put you on the wrong track.